### PR TITLE
#164802372 Filter asset logs

### PR DESCRIPTION
**Pull request by 3N61N33R**
#### What does this PR do?
Filters asset logs by the user, day, month and year

#### Description of Task to be completed?
Given I am a user with access
When I filter logs by either the user, day, month and year.
Then I should see the filtered data

#### How should this be manually tested?
- fetch the branch `git fetch origin ft-filter-asset-logs-164802372`
- checkout to the branch `git checkout ft-filter-asset-logs-164802372`
- visit the endpoints: 
   `http://127.0.0.1:8000/api/v1/asset-logs?user=<email>`
   `http://127.0.0.1:8000/api/v1/asset-logs?day=<day>`
   `http://127.0.0.1:8000/api/v1/asset-logs?month=<month>`
   `http://127.0.0.1:8000/api/v1/asset-logs?year=<year>`

#### What are the relevant pivotal tracker stories?
[#164802372](https://www.pivotaltracker.com/story/show/164802372)

### CODE REVIEW

**Review by kmazi — CHANGES_REQUESTED** [review]
You should add integration tests, at least, to your feature.

File: api/filters.py [hunk 1]
```diff
@@ -12,99 +12,115 @@
 
 logger = logging.getLogger(__name__)
 
-NULL_VALUE = 'unspecified'
+NULL_VALUE = "unspecified"
```
Comment (RonKbS) [inline api/filters.py]: From the demo we had earlier, I believe this has been properly implemented. 
However, the changes you've included in relation to formatting are also in this PR -> `https://github.com/AndelaOSP/art-backend/pull/292`. I'd recommend you exclude them from this PR. Merge conflicts are also likely to occur once the other one is merged.

File: api/filters.py [hunk 2]
```diff
@@ -12,99 +12,115 @@
...
-        field_name='assigned_to__user__email', lookup_expr='icontains'
+        field_name="assigned_to__user__email", lookup_expr="icontains"
     )
     model_number = filters.CharFilter(
-        field_name='model_number__name',
-        lookup_expr='iexact',
-        method='filter_contains_with_multiple_query_values',
+        field_name="model_number__name",
+        lookup_expr="iexact",
+        method="filter_contains_with_multiple_query_values",
```
Comment (kmazi) [inline api/filters.py:51]: Your lookup is exact yet you are using a filter for `icontains`. There seems to be no need for the multiple query value filter here.

File: api/filters.py [hunk 3]
```diff
@@ -12,99 +12,115 @@
...
     )
     serial_number = filters.CharFilter(
-        field_name='serial_number',
-        lookup_expr='icontains',
-        method='filter_contains_with_multiple_query_values',
+        field_name="serial_number",
+        lookup_expr="icontains",
+        method="filter_contains_with_multiple_query_values",
```
Comment (kmazi) [inline api/filters.py]: I don't think this should be used here. I also think querying for serial_number should be exact. @egichuri what do you reckon?

File: api/filters.py [hunk 4]
```diff
@@ -12,99 +12,115 @@
...
     )
     asset_type = filters.CharFilter(
-        field_name='model_number__asset_make__asset_type__name',
-        lookup_expr='iexact',
-        method='filter_contains_with_multiple_query_values',
+        field_name="model_number__asset_make__asset_type__name",
+        lookup_expr="iexact",
```
Comment (kmazi) [inline api/filters.py]: This is better as `icontain` so that a user can get useful response values even when not sure of the asset type name or it's spelling. @egichuri

File: api/filters.py (hunk 1 above)
Comment (3N61N33R) [reply api/filters.py]: @RonKbS this is a file change by black

File: api/filters.py (hunk 2 above)
Comment (3N61N33R) [reply api/filters.py:51]: hey @kmazi I did not work on this, most of these are highlighted because of a reformatting by `black`
but I could check them out later on

File: api/filters.py (hunk 1 above)
Comment (RonKbS) [reply api/filters.py]: Yes it is. However, there was a PR that's been merged. It was meant to deal with the formatting issues. => `https://github.com/AndelaOSP/art-backend/pull/292`. You'll probably need to look into undoing the current changes as well. You have 65 files worked on now.

**Review by kmazi — CHANGES_REQUESTED** [review]
You ought to assert that a user can filter asset-logs by the asset assignee(user), year, month and day. Your tests should also assert the edge cases that could be encountered while filtering.

**Comment by egichuri** [comment]
@graycadeau ~please merge `develop` to your branch.~ had you branched off master or develop?

File: api/tests/test_asset_log.py [hunk 5]
```diff
@@ -337,6 +338,109 @@ def test_authenticated_admin_user_get_of_asset_logs_invalid_filter(
...
         self.assertEqual(len(response.data["results"]), 0)
 
+    @patch("api.authentication.auth.verify_id_token")
+    def test_filter_asset_logs_by_year(self, mock_verify_id_token):
+        mock_verify_id_token.return_value = {"email": self.admin_user.email}
+        AssetLog.objects.create(
+            checked_by=self.security_user, asset=self.test_other_asset, log_type=CHECKIN
+        )
+        asset_logs_url = (
+            f"{self.asset_logs_url}?year={self.test_other_asset.created_at.year}"
```
Comment (kmazi) [inline api/tests/test_asset_log.py]: I believe this filter is for the year asset-logs were created and not when the assets they log were created.
You can update the `created_at` attribute of the created log to set a custom year for your assertion. Check out [this](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#django.db.models.query.QuerySet.update).

File: api/tests/test_asset_log.py [hunk 6]
```diff
@@ -337,6 +338,109 @@ def test_authenticated_admin_user_get_of_asset_logs_invalid_filter(
...
+
+        response = client.get(
+            asset_logs_url, HTTP_AUTHORIZATION=f"Token {self.token_admin}"
+        )
+        data = response.data["results"]
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(
+            data[0]["created_at"].split("-")[0],
+            f"{self.test_other_asset.created_at.year}",
+        )
```
Comment (kmazi) [inline api/tests/test_asset_log.py:361]: Let your assertion show the number of items filtered by the query. You can do the same for other assertions.

File: api/tests/test_asset_log.py [hunk 7]
```diff
@@ -337,6 +338,109 @@ def test_authenticated_admin_user_get_of_asset_logs_invalid_filter(
...
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(len(response.data["results"]), 0)
+
+    @patch("api.authentication.auth.verify_id_token")
+    def test_filter_asset_logs_by_month(self, mock_verify_id_token):
+        mock_verify_id_token.return_value = {"email": self.admin_user.email}
+        AssetLog.objects.create(
+            checked_by=self.security_user, asset=self.test_other_asset, log_type=CHECKIN
+        )
```
Comment (kmazi) [inline api/tests/test_asset_log.py:391]: You could update the month this asset-log was created and assert that you could filter it out from the rest. Just update the `created_at` attribute.

File: api/tests/test_asset_log.py [hunk 8]
```diff
@@ -5,6 +5,7 @@
 from django.contrib.auth import get_user_model
 from django.core.exceptions import ValidationError
 from rest_framework.test import APIClient
+from datetime import datetime
```
Comment (kmazi) [inline api/tests/test_asset_log.py]: It's standard practice to import datetime as dt. You could do this `from datetime import datetime as dt`.

**Review by kmazi — CHANGES_REQUESTED** [review]
It's great that you've tested for edge cases. Update the created_at attribute of the logs you created in your test methods to a value you can assert for. You could also let your tests show the number of filtered items for both valid and edge cases.

You seem to have commits other than yours. Did you branch of develop at the start of your task? If you didn't you could rebase onto develop using the `git rebase --onto <new-base-branch> <old-base-branch> <commit-head>`.

You could rebase interactively to select out the other commits you know.

File: api/tests/test_asset_log.py (hunk 8 above)
Comment (egichuri) [reply api/tests/test_asset_log.py]: wouldn't call it standard practice.. i'd say it applies when you're importing other objects from the datetime module

File: api/tests/test_asset_log.py (hunk 7 above)
Comment (3N61N33R) [reply api/tests/test_asset_log.py:391]: I tried this out and it brought errors that's why I left it out

**Comment by 3N61N33R** [comment]
> @graycadeau ~please merge `develop` to your branch.~ had you branched off master or develop?

@egichuri I branched off develop

File: api/tests/test_asset_log.py (hunk 6 above)
Comment (3N61N33R) [reply api/tests/test_asset_log.py:361]: @kmazi Wouldn't this mean that I first have to show the number of items that are there before the filter?
All the dates are autogenerated and get (datetime.now()) each time you run the test
I tried updating the `created_at` time and it would resolve to a zero everytime.

**Comment by 3N61N33R** [comment]
> It's great that you've tested for edge cases. Update the created_at attribute of the logs you created in your test methods to a value you can assert for. You could also let your tests show the number of filtered items for both valid and edge cases.
> 
> You seem to have commits other than yours. Did you branch of develop at the start of your task? If you didn't you could rebase onto develop using the `git rebase --onto <new-base-branch> <old-base-branch> <commit-head>`.
> 
> You could rebase interactively to select out the other commits you know.

@kmazi commits came after rebasing and I could not force push, that is why I have the commits
I could drop the commits that have already been merged to develop and remain with my commits

File: api/tests/test_asset_log.py [hunk 9]
```diff
@@ -316,6 +338,126 @@ def test_authenticated_admin_user_get_of_asset_logs_invalid_filter(
...
+        )
+        asset_logs_url = f"{self.asset_logs_url}?day=456"
+
+        response = client.get(
+            asset_logs_url, HTTP_AUTHORIZATION=f"Token {self.token_admin}"
+        )
+        self.assertEqual(response.status_code, 200)
+        logs_count = AssetLog.objects.filter(created_at__day="456").count()
+        self.assertEqual(len(response.data["results"]), logs_count)
+
```
Comment (egichuri) [inline api/tests/test_asset_log.py:460]: one last thing 😅  
could we have one more test checking for the 3 attribute in the same fetch? eg `year=2019&month=5&day=15`

File: settings/base.py [hunk 10]
```diff
@@ -139,6 +139,7 @@
         "rest_framework.authentication.SessionAuthentication",
         "oauth2_provider.contrib.rest_framework.OAuth2Authentication",
     ),
+    "DEFAULT_FILTER_BACKENDS": ("django_filters.rest_framework.DjangoFilterBackend",),
```
Comment (egichuri) [inline settings/base.py:142]: now that we have this we can remove all traces of `filter_backends = (filters.DjangoFilterBackend,)` in the views.

**Review by egichuri — CHANGES_REQUESTED** [review]

File: api/tests/test_asset_log.py [hunk 11]
```diff
@@ -316,6 +338,126 @@ def test_authenticated_admin_user_get_of_asset_logs_invalid_filter(
         self.assertEqual(response.status_code, 200)
         self.assertEqual(len(response.data["results"]), 0)
 
+    @patch(
+        "django.utils.timezone.now",
+        return_value=datetime(2017, 9, 22, 17, 1, 26, 842_150),
```
Comment (kmazi) [inline api/tests/test_asset_log.py:343]: It's really nice you mocked it out.

**Review by kmazi — APPROVED** [review]

**Review by egichuri — APPROVED** [review]
